File: sdk-python/obsvr/strict_receipt_coordinator_support.py
"""Validation and canonical helpers for strict receipt coordination."""

from __future__ import annotations

import copy
import hashlib
from typing import Any, Callable, Dict, List

from .action_context import build_action_context
from .strict_receipt import (
    STRICT_RECEIPT_PROFILE_VERSION,
    STRICT_RECEIPT_SCHEMA,
    sign_strict_receipt,
    strict_receipt_key_id,
)
from .tool_pinning import _canonical_json_for_hash

_HEX = frozenset("0123456789abcdef")
_MAX_SAFE_INTEGER = 9_007_199_254_740_991


def coordinator_text(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{field} must be a nonblank string")
    return value


def coordinator_hash(value: Any, field: str) -> str:
    if (
        not isinstance(value, str)
        or len(value) != 64
        or any(char not in _HEX for char in value)
    ):
        raise ValueError(f"{field} must be 64 lowercase hex characters")
    return value


def normalized_strings(value: Any, field: str) -> List[str]:
    if not isinstance(value, list):
        raise ValueError(f"{field} must be an array")
    values = [
        coordinator_text(item, f"{field}[{index}]")
        for index, item in enumerate(value)
    ]
    return sorted(set(values), key=lambda item: tuple(ord(char) for char in item))


def safe_integer(value: Any, field: str) -> int:
    if (
        isinstance(value, bool)
        or not isinstance(value, int)
        or value < 0
        or value > _MAX_SAFE_INTEGER
    ):
        raise ValueError(f"{field} must be a nonnegative safe integer")
    return value


def safe_add(left: int, right: int) -> int:
    result = left + right
    if result > _MAX_SAFE_INTEGER:
        raise ValueError("value exceeds safe integer range")
    return result


def build_coordinator_context(
    context: Dict[str, Any], session_id: str, prior_actions: List[Dict[str, Any]]
) -> Dict[str, Any]:
    if not isinstance(context, dict):
        raise ValueError("context must be an object")
    if "prior_actions" in context:
        raise ValueError("caller prior_actions are not accepted")
    if "session_id" in context:
        raise ValueError("caller session_id is not accepted")
    return build_action_context(
        {
            **copy.deepcopy(context),
            "session_id": session_id,
            "prior_actions": copy.deepcopy(prior_actions),
        }
    )


def context_input_from_document(context: Dict[str, Any]) -> Dict[str, Any]:
    result: Dict[str, Any] = {
        "agent_id": context["agent"]["agent_id"],
        "active_intents": list(context["agent"]["active_intents"]),
        "current_action": copy.deepcopy(context["action"]),
        "run_id": context["run_id"],
    }
    if "role" in context["agent"]:
        result["agent_role"] = context["agent"]["role"]
    if "privilege_scope" in context["agent"]:
        result["privilege_scope"] = list(context["agent"]["privilege_scope"])
    if "thread_id" in context:
        result["thread_id"] = context["thread_id"]
    return result


def request_fingerprint(
    *, context: Dict[str, Any], base_result: Dict[str, Any],
    policy_version: str, rule_ids: List[str], session_id: str,
    action_id: str,
) -> str:
    normalized_context = build_action_context(
        {**context, "session_id": session_id, "prior_actions": []}
    )
    document = {
        "schema": "obsvr-strict-decision-request-v1",
        "action_id": coordinator_text(action_id, "action_id"),
        "context": normalized_context,
        "base_result": dict(base_result),
        "policy_version": coordinator_text(policy_version, "policy_version"),
        "rule_ids": normalized_strings(rule_ids, "rule_ids"),
    }
    return hashlib.sha256(
        _canonical_json_for_hash(document).encode("utf-8")
    ).hexdigest()


def trusted_approval_result(
    value: Any, expected: Dict[str, Any], suspension_expiry: int
) -> Dict[str, str]:
    if not isinstance(value, dict):
        raise ValueError("approval verifier returned an invalid result")
    required = {
        "request_id", "action_hash", "principal_id", "decision",
        "source_hash", "expires_at_ms",
    }
    allowed = required | {"principal_ref_hash"}
    if not required.issubset(value) or not set(value).issubset(allowed):
        raise ValueError("approval verifier returned an invalid result")
    request_id = coordinator_text(value.get("request_id"), "trusted request_id")
    action_hash = coordinator_hash(value.get("action_hash"), "trusted action_hash")
    principal_id = coordinator_text(value.get("principal_id"), "trusted principal_id")
    principal_ref_hash = (
        None
        if "principal_ref_hash" not in value
        else coordinator_hash(
            value.get("principal_ref_hash"), "trusted principal_ref_hash"
        )
    )
    source_hash = coordinator_hash(value.get("source_hash"), "trusted source_hash")
    expires_at = safe_integer(value.get("expires_at_ms"), "trusted expires_at_ms")
    if (
        request_id != expected["request_id"]
        or action_hash != expected["action_hash"]
        or value.get("decision") != expected["decision"]
    ):
        raise ValueError("approval verifier result does not match expected binding")
    if expires_at > suspension_expiry:
        raise ValueError("approval verifier expiry exceeds suspension expiry")
    if expected["decision"] == "granted" and (
        expected["current_time_ms"] >= expires_at
        or expected["current_time_ms"] >= suspension_expiry
    ):
        raise ValueError("trusted approval is expired")
    result = {
        "request_id": request_id,
        "action_hash": action_hash,
        "principal_id": principal_id,
        "source_hash": source_hash,
    }
    if principal_ref_hash is not None:
        result["principal_ref_hash"] = principal_ref_hash
    return result


def _projection(context: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "agent_id": context["agent"]["agent_id"],
        "role": context["agent"].get("role"),
        "privilege_scope": context["agent"].get("privilege_scope"),
        "active_intents": context["agent"]["active_intents"],
        "action": dict(context["action"]),
        "run_id": context["run_id"],
        "session_id": context.get("session_id"),
        "thread_id": context.get("thread_id"),
    }


def validate_deferred_changes(
    *, original_context: Dict[str, Any], current_context: Dict[str, Any],
    original_base: Dict[str, Any], current_base: Dict[str, Any],
    required_fields: List[str]
) -> None:
    required = set(required_fields)
    before = _projection(original_context)
    after = _projection(current_context)
    if "action.target" in required:
        before["action"].pop("target", None)
        after["action"].pop("target", None)
    if "active_intents" in required:
        before.pop("active_intents")
        after.pop("active_intents")
    if _canonical_json_for_hash(before) != _canonical_json_for_hash(after):
        raise ValueError("DEFER resolution changed fields outside required_fields")
    if "policy_evaluation" in required:
        return
    allowed_base = {
        field for field in required_fields
        if field == "modified_arguments_hash" or field.startswith("approval_")
    }
    old_base = dict(original_base)
    new_base = dict(current_base)
    for field in allowed_base:
        old_base.pop(field, None)
        new_base.pop(field, None)
    if _canonical_json_for_hash(old_base) != _canonical_json_for_hash(new_base):
        raise ValueError(
            "DEFER resolution changed base result outside required_fields"
        )


def sign_coordinator_resolution(
    *, prior: Dict[str, Any], evaluation: Dict[str, Any],
    context: Dict[str, Any], base_result: Dict[str, Any], policy_version: str,
    rule_ids: List[str], method: str, principal_id: str, source_hash: str,
    timestamp: int, clamped: bool, sequence: int, session_id: str,
    previous_hash: str | None, sdk_version: str, signer: Any,
    include_public_key: bool,
) -> Dict[str, Any]:
    action = {
        "action_id": prior["body"]["action"]["action_id"],
        "kind": context["action"]["kind"],
        "name": context["action"]["name"],
        "arguments_hash": context["action"]["arguments_hash"],
    }
    if "target" in context["action"]:
        action["target"] = context["action"]["target"]
    if evaluation["outcome"] == "MODIFY":
        action["effective_arguments_hash"] = base_result["modified_arguments_hash"]
    body = {
        "schema": STRICT_RECEIPT_SCHEMA,
        "profile_version": STRICT_RECEIPT_PROFILE_VERSION,
        "record_type": "resolution",
        "receipt_id": f"{session_id}:{sequence}",
        "session_id": session_id,
        "sequence": sequence,
        "timestamp_ms": timestamp,
        "clock_regression_clamped": clamped,
        "previous_receipt_hash": previous_hash,
        "sdk": {"language": "python", "version": sdk_version},
        "initiator": copy.deepcopy(prior["body"]["initiator"]),
        "action": action,
        "context": {
            "schema": "obsvr-action-context-v1",
            "context_hash": evaluation["context_hash"],
            "run_id": context["run_id"],
        },
        "evaluation": {
            "input_hash": evaluation["input_hash"],
            "policy_hash": evaluation["policy_hash"],
            "evaluator_hash": evaluation["evaluator_hash"],
            "engine_version": evaluation["engine_version"],
            "policy_version": coordinator_text(policy_version, "policy_version"),
            "outcome": evaluation["outcome"],
            "reason_code": f"resolution_{method}",
            "rule_ids": normalized_strings(rule_ids, "rule_ids"),
        },
        "execution_authorized": evaluation["outcome"] in ("ALLOW", "MODIFY"),
        "resolution": {
            "resolves_receipt_hash": prior["receipt_hash"],
            "suspension_id": prior["body"]["suspension"]["suspension_id"],
            "method": method,
            "resolver_principal_id": principal_id,
            "resolution_source_hash": source_hash,
            "resolved_at_ms": timestamp,
        },
    }
    if "thread_id" in context:
        body["context"]["thread_id"] = context["thread_id"]
    return sign_strict_receipt(body, signer, include_public_key)


def sign_coordinator_decision(
    *, action_id: str, context: Dict[str, Any], base_result: Dict[str, Any],
    evaluation: Dict[str, Any], policy_version: str, rule_ids: List[str],
    timestamp: int, clamped: bool, sequence: int, session_id: str,
    previous_hash: str | None, sdk_version: str, signer: Any,
    include_public_key: bool, defer_ttl_ms: int,
    approval_request_pending: Callable[[str], bool],
) -> Dict[str, Any]:
    body: Dict[str, Any] = {
        "schema": STRICT_RECEIPT_SCHEMA,
        "profile_version": STRICT_RECEIPT_PROFILE_VERSION,
        "record_type": "decision",
        "receipt_id": f"{session_id}:{sequence}",
        "session_id": session_id, "sequence": sequence,
        "timestamp_ms": timestamp, "clock_regression_clamped": clamped,
        "previous_receipt_hash": previous_hash,
        "sdk": {"language": "python", "version": sdk_version},
        "initiator": {
            "agent_id": context["agent"]["agent_id"],
            "key_id": strict_receipt_key_id(signer.raw_public_key),
        },
        "action": {
            "action_id": action_id, "kind": context["action"]["kind"],
            "name": context["action"]["name"],
            "arguments_hash": context["action"]["arguments_hash"],
        },
        "context": {
            "schema": "obsvr-action-context-v1",
            "context_hash": evaluation["context_hash"],
            "run_id": context["run_id"],
        },
        "evaluation": {
            "input_hash": evaluation["input_hash"],
            "policy_hash": evaluation["policy_hash"],
            "evaluator_hash": evaluation["evaluator_hash"],
            "engine_version": evaluation["engine_version"],
            "policy_version": coordinator_text(policy_version, "policy_version"),
            "outcome": evaluation["outcome"],
            "reason_code": evaluation["reason_code"],
            "rule_ids": normalized_strings(rule_ids, "rule_ids"),
        },
        "execution_authorized": evaluation["outcome"] in ("ALLOW", "MODIFY"),
    }
    if "target" in context["action"]:
        body["action"]["target"] = context["action"]["target"]
    if "thread_id" in context:
        body["context"]["thread_id"] = context["thread_id"]
    if evaluation["outcome"] == "MODIFY":
        body["action"]["effective_arguments_hash"] = base_result[
            "modified_arguments_hash"
        ]
    if evaluation["outcome"] == "STEP_UP":
        request_id = base_result["approval_request_id"]
        if approval_request_pending(request_id):
            raise ValueError("approval_request_id is already pending")
        body["suspension"] = {
            "suspension_id": request_id, "type": "approval",
            "status": "pending", "required_fields": [],
            "expires_at_ms": base_result["approval_expires_at_ms"],
            "approval_request_id": request_id,
            "approval_action_hash": base_result["approval_action_hash"],
        }
        if body["suspension"]["expires_at_ms"] <= timestamp:
            raise ValueError("approval expiry must follow decision timestamp")
    elif evaluation["outcome"] == "DEFER":
        body["suspension"] = {
            "suspension_id": f"defer:{session_id}:{sequence}",
            "type": "context", "status": "pending",
            "required_fields": evaluation["required_fields"],
            "expires_at_ms": safe_add(timestamp, defer_ttl_ms),
        }
    return sign_strict_receipt(body, signer, include_public_key)


def validate_coordinator_resolution(
    *, pending: Dict[str, Any], method: str, context: Dict[str, Any],
    base_result: Dict[str, Any], evaluation: Dict[str, Any], resolved_at: int,
    resolver_principal_id: Any, resolution_source_hash: Any,
    approval_evidence_value: Any, approval_verifier: Callable[[Any, Any], Any],
    approval_request_receipt: Callable[[str], Any],
) -> tuple[str, str]:
    prior = pending["receipt"]
    suspension = prior["body"]["suspension"]
    if evaluation["outcome"] not in {"ALLOW", "DENY", "MODIFY"}:
        raise ValueError("resolution evaluation must be final")
    action = context["action"]
    prior_action = prior["body"]["action"]
    if (
        context["agent"]["agent_id"] != prior["body"]["initiator"]["agent_id"]
        or any(action.get(key) != prior_action.get(key)
               for key in ("kind", "name", "arguments_hash", "target"))
    ):
        raise ValueError("resolution action or initiator does not match suspension")
    approval_method = method in {"approval_granted", "approval_denied"}
    if (
        (suspension["type"] == "approval") != approval_method
        and method not in ("expired", "cancelled")
    ):
        raise ValueError("resolution method does not match suspension")
    if approval_method:
        expected = {
            "request_id": suspension["approval_request_id"],
            "action_hash": suspension["approval_action_hash"],
            "decision": "granted" if method == "approval_granted" else "denied",
            "current_time_ms": resolved_at,
        }
        evidence = trusted_approval_result(
            approval_verifier(approval_evidence_value, expected),
            expected, suspension["expires_at_ms"],
        )
        if approval_request_receipt(evidence["request_id"]) != prior["receipt_hash"]:
            raise ValueError("approval request belongs to another suspension")
        if resolver_principal_id is not None or resolution_source_hash is not None:
            raise ValueError("approval source must come from approval evidence")
        principal, source = evidence["principal_id"], evidence["source_hash"]
    else:
        if approval_evidence_value is not None:
            raise ValueError("approval evidence requires an approval method")
        if suspension["type"] == "context":
            validate_deferred_changes(
                original_context=pending["context"], current_context=context,
                original_base=pending["base_result"], current_base=base_result,
                required_fields=suspension["required_fields"],
            )
        principal = coordinator_text(resolver_principal_id, "resolver_principal_id")
        source = coordinator_hash(resolution_source_hash, "resolution_source_hash")
    if method in ("approval_granted", "context_supplied") and (
        resolved_at >= suspension["expires_at_ms"]
    ):
        raise ValueError("authorization cannot occur after suspension expiry")
    if method in {"approval_denied", "expired", "cancelled"} and (
        evaluation["outcome"] != "DENY"
    ):
        raise ValueError("resolution method requires DENY")
    if method == "approval_granted" and evaluation["outcome"] == "DENY":
        raise ValueError("approval_granted requires ALLOW or MODIFY")
    if evaluation["outcome"] == "MODIFY" and (
        coordinator_hash(base_result.get("modified_arguments_hash"),
                         "modified_arguments_hash")
        == prior_action["arguments_hash"]
    ):
        raise ValueError("effective_arguments_hash must change arguments")
    return principal, source

File: sdk-typescript/src/governance/strict-receipt-coordinator-support.ts
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import { sha256Hex } from '../policy/decision-record.js';
import type { IntentBaseResult } from '../policy/intent-alignment.js';
import type { IntentAlignmentResult } from '../policy/intent-alignment.js';
import type { DeviceSigner } from '../proxy/device-identity.js';
import {
  buildActionContext,
  type ActionContextDocument,
  type ActionContextInput,
  type PriorActionInput,
} from './action-context.js';
import {
  STRICT_RECEIPT_PROFILE_VERSION,
  STRICT_RECEIPT_SCHEMA,
  signStrictReceipt,
  strictReceiptKeyId,
  type StrictReceiptBody,
  type StrictReceiptEnvelope,
} from './strict-receipt.js';

const HEX64 = /^[0-9a-f]{64}$/;

export interface StrictApprovalExpectation {
  request_id: string;
  action_hash: string;
  decision: 'granted' | 'denied';
  current_time_ms: number;
}

export interface TrustedApprovalResult {
  request_id: string;
  action_hash: string;
  principal_id: string;
  /** Pseudonymous stable identity used for separation-of-duties checks. */
  principal_ref_hash?: string;
  decision: 'granted' | 'denied';
  source_hash: string;
  expires_at_ms: number;
}

export type StrictApprovalVerifier = (
  evidence: unknown, expected: StrictApprovalExpectation,
) => TrustedApprovalResult;

export function coordinatorText(value: unknown, field: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new Error(`${field} must be a nonblank string`);
  }
  return value;
}

export function coordinatorHash(value: unknown, field: string): string {
  if (typeof value !== 'string' || !HEX64.test(value)) {
    throw new Error(`${field} must be 64 lowercase hex characters`);
  }
  return value;
}

export function coordinatorSafeInteger(value: unknown, field: string): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) {
    throw new Error(`${field} must be a nonnegative safe integer`);
  }
  return value;
}

export function cloneCoordinatorValue<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

function compareScalars(left: string, right: string): number {
  const a = Array.from(left, (char) => char.codePointAt(0) as number);
  const b = Array.from(right, (char) => char.codePointAt(0) as number);
  for (let index = 0; index < Math.min(a.length, b.length); index++) {
    if (a[index] !== b[index]) return a[index] - b[index];
  }
  return a.length - b.length;
}

export function normalizedStrings(value: unknown, field: string): string[] {
  if (!Array.isArray(value)) throw new Error(`${field} must be an array`);
  const values = value.map((item, index) => coordinatorText(item, `${field}[${index}]`));
  values.sort(compareScalars);
  return values.filter((item, index) => index === 0 || item !== values[index - 1]);
}

export function canonicalHash(value: unknown): string {
  return sha256Hex(canonicalJsonForHash(value));
}

export function contextInputFromDocument(
  context: ActionContextDocument,
): Omit<ActionContextInput, 'prior_actions' | 'session_id'> {
  return {
    agent_id: context.agent.agent_id,
    active_intents: [...context.agent.active_intents],
    ...(context.agent.role === undefined ? {} : { agent_role: context.agent.role }),
    ...(context.agent.privilege_scope === undefined
      ? {} : { privilege_scope: [...context.agent.privilege_scope] }),
    current_action: { ...context.action },
    run_id: context.run_id,
    ...(context.thread_id === undefined ? {} : { thread_id: context.thread_id }),
  };
}

export function buildCoordinatorContext(
  input: Omit<ActionContextInput, 'prior_actions' | 'session_id'>,
  sessionId: string,
  priorActions: PriorActionInput[],
): ActionContextDocument {
  const raw = input as unknown as Record<string, unknown>;
  if (Object.prototype.hasOwnProperty.call(raw, 'prior_actions')) {
    throw new Error('caller prior_actions are not accepted');
  }
  if (Object.prototype.hasOwnProperty.call(raw, 'session_id')) {
    throw new Error('caller session_id is not accepted');
  }
  return buildActionContext({
    ...input,
    session_id: sessionId,
    prior_actions: priorActions.map((item) => ({
      ...item, data_classifications: [...item.data_classifications],
    })),
  });
}

export function addSafeIntegers(left: number, right: number): number {
  const result = left + right;
  if (!Number.isSafeInteger(result)) throw new Error('value exceeds safe integer range');
  return result;
}

export function signCoordinatorResolution(params: {
  prior: StrictReceiptEnvelope;
  evaluation: IntentAlignmentResult;
  context: ActionContextDocument;
  base_result: IntentBaseResult;
  policy_version: string;
  rule_ids: string[];
  method: NonNullable<StrictReceiptBody['resolution']>['method'];
  principal_id: string;
  source_hash: string;
  timestamp: number;
  clamped: boolean;
  sequence: number;
  session_id: string;
  previous_hash: string | null;
  sdk_version: string;
  signer: DeviceSigner;
  include_public_key: boolean;
}): StrictReceiptEnvelope {
  const action: StrictReceiptBody['action'] = {
    action_id: params.prior.body.action.action_id,
    kind: params.context.action.kind,
    name: params.context.action.name,
    arguments_hash: params.context.action.arguments_hash,
  };
  if (params.context.action.target !== undefined) action.target = params.context.action.target;
  if (params.evaluation.outcome === 'MODIFY') {
    action.effective_arguments_hash = params.base_result.modified_arguments_hash;
  }
  const body: StrictReceiptBody = {
    schema: STRICT_RECEIPT_SCHEMA,
    profile_version: STRICT_RECEIPT_PROFILE_VERSION,
    record_type: 'resolution',
    receipt_id: `${params.session_id}:${params.sequence}`,
    session_id: params.session_id,
    sequence: params.sequence,
    timestamp_ms: params.timestamp,
    clock_regression_clamped: params.clamped,
    previous_receipt_hash: params.previous_hash,
    sdk: { language: 'typescript', version: params.sdk_version },
    initiator: { ...params.prior.body.initiator },
    action,
    context: {
      schema: 'obsvr-action-context-v1',
      context_hash: params.evaluation.context_hash,
      run_id: params.context.run_id,
      ...(params.context.thread_id === undefined ? {} : { thread_id: params.context.thread_id }),
    },
    evaluation: {
      input_hash: params.evaluation.input_hash,
      policy_hash: params.evaluation.policy_hash,
      evaluator_hash: params.evaluation.evaluator_hash,
      engine_version: params.evaluation.engine_version,
      policy_version: params.policy_version,
      outcome: params.evaluation.outcome,
      reason_code: `resolution_${params.method}`,
      rule_ids: params.rule_ids,
    },
    execution_authorized: params.evaluation.outcome === 'ALLOW'
      || params.evaluation.outcome === 'MODIFY',
    resolution: {
      resolves_receipt_hash: params.prior.receipt_hash,
      suspension_id: params.prior.body.suspension!.suspension_id,
      method: params.method,
      resolver_principal_id: params.principal_id,
      resolution_source_hash: params.source_hash,
      resolved_at_ms: params.timestamp,
    },
  };
  return signStrictReceipt(body, params.signer, params.include_public_key);
}

export function signCoordinatorDecision(params: {
  action_id: string; context: ActionContextDocument;
  base_result: IntentBaseResult; evaluation: IntentAlignmentResult;
  policy_version: string; rule_ids: string[]; timestamp: number; clamped: boolean;
  sequence: number; session_id: string; previous_hash: string | null;
  sdk_version: string; signer: DeviceSigner; include_public_key: boolean;
  defer_ttl_ms: number; approval_request_pending: (requestId: string) => boolean;
}): StrictReceiptEnvelope {
  const body: StrictReceiptBody = {
    schema: STRICT_RECEIPT_SCHEMA,
    profile_version: STRICT_RECEIPT_PROFILE_VERSION,
    record_type: 'decision',
    receipt_id: `${params.session_id}:${params.sequence}`,
    session_id: params.session_id,
    sequence: params.sequence,
    timestamp_ms: params.timestamp,
    clock_regression_clamped: params.clamped,
    previous_receipt_hash: params.previous_hash,
    sdk: { language: 'typescript', version: params.sdk_version },
    initiator: {
      agent_id: params.context.agent.agent_id,
      key_id: strictReceiptKeyId(params.signer.rawPublicKey),
    },
    action: {
      action_id: params.action_id, kind: params.context.action.kind,
      name: params.context.action.name,
      arguments_hash: params.context.action.arguments_hash,
    },
    context: {
      schema: 'obsvr-action-context-v1', context_hash: params.evaluation.context_hash,
      run_id: params.context.run_id,
    },
    evaluation: {
      input_hash: params.evaluation.input_hash,
      policy_hash: params.evaluation.policy_hash,
      evaluator_hash: params.evaluation.evaluator_hash,
      engine_version: params.evaluation.engine_version,
      policy_version: params.policy_version,
      outcome: params.evaluation.outcome,
      reason_code: params.evaluation.reason_code,
      rule_ids: params.rule_ids,
    },
    execution_authorized: params.evaluation.outcome === 'ALLOW'
      || params.evaluation.outcome === 'MODIFY',
  };
  if (params.context.action.target !== undefined) {
    body.action.target = params.context.action.target;
  }
  if (params.context.thread_id !== undefined) body.context.thread_id = params.context.thread_id;
  if (params.evaluation.outcome === 'MODIFY') {
    body.action.effective_arguments_hash = params.base_result.modified_arguments_hash;
  }
  if (params.evaluation.outcome === 'STEP_UP') {
    const requestId = params.base_result.approval_request_id as string;
    if (params.approval_request_pending(requestId)) {
      throw new Error('approval_request_id is already pending');
    }
    body.suspension = {
      suspension_id: requestId, type: 'approval', status: 'pending', required_fields: [],
      expires_at_ms: params.base_result.approval_expires_at_ms as number,
      approval_request_id: requestId,
      approval_action_hash: params.base_result.approval_action_hash,
    };
    if (body.suspension.expires_at_ms <= params.timestamp) {
      throw new Error('approval expiry must follow decision timestamp');
    }
  } else if (params.evaluation.outcome === 'DEFER') {
    body.suspension = {
      suspension_id: `defer:${params.session_id}:${params.sequence}`,
      type: 'context', status: 'pending',
      required_fields: params.evaluation.required_fields as string[],
      expires_at_ms: addSafeIntegers(params.timestamp, params.defer_ttl_ms),
    };
  }
  return signStrictReceipt(body, params.signer, params.include_public_key);
}

export function requestFingerprint(params: {
  context: Omit<ActionContextInput, 'prior_actions' | 'session_id'>;
  base_result: IntentBaseResult;
  policy_version: string;
  rule_ids: string[];
  session_id: string;
  action_id: string;
}): string {
  const context = buildActionContext({
    ...params.context,
    session_id: params.session_id,
    prior_actions: [],
  });
  return sha256Hex(canonicalJsonForHash({
    schema: 'obsvr-strict-decision-request-v1',
    action_id: coordinatorText(params.action_id, 'action_id'),
    context,
    base_result: { ...params.base_result },
    policy_version: coordinatorText(params.policy_version, 'policy_version'),
    rule_ids: normalizedStrings(params.rule_ids, 'rule_ids'),
  }));
}

export function trustedApprovalResult(
  value: unknown,
  expected: StrictApprovalExpectation,
  suspensionExpiry: number,
): {
  requestId: string;
  actionHash: string;
  principalId: string;
  principalRefHash?: string;
  sourceHash: string;
} {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('approval verifier returned an invalid result');
  }
  const raw = value as Record<string, unknown>;
  const required = new Set([
    'request_id', 'action_hash', 'principal_id', 'decision',
    'source_hash', 'expires_at_ms',
  ]);
  const allowed = new Set([...required, 'principal_ref_hash']);
  const unknown = Object.keys(raw).filter((key) => !allowed.has(key));
  const missing = [...required].filter((key) => !Object.prototype.hasOwnProperty.call(raw, key));
  if (unknown.length > 0 || missing.length > 0) {
    throw new Error('approval verifier returned an invalid result');
  }
  const requestId = coordinatorText(raw.request_id, 'trusted approval request_id');
  const actionHash = coordinatorHash(raw.action_hash, 'trusted approval action_hash');
  const principalId = coordinatorText(raw.principal_id, 'trusted approval principal_id');
  const principalRefHash = raw.principal_ref_hash === undefined
    ? undefined
    : coordinatorHash(raw.principal_ref_hash, 'trusted approval principal_ref_hash');
  const sourceHash = coordinatorHash(raw.source_hash, 'trusted approval source_hash');
  const expiresAt = coordinatorSafeInteger(raw.expires_at_ms, 'trusted approval expires_at_ms');
  if (requestId !== expected.request_id || actionHash !== expected.action_hash
    || raw.decision !== expected.decision) {
    throw new Error('approval verifier result does not match expected binding');
  }
  if (expiresAt > suspensionExpiry) {
    throw new Error('approval verifier expiry exceeds suspension expiry');
  }
  if (expected.decision === 'granted'
    && (expected.current_time_ms >= expiresAt
      || expected.current_time_ms >= suspensionExpiry)) {
    throw new Error('trusted approval is expired');
  }
  return {
    requestId, actionHash, principalId,
    ...(principalRefHash === undefined ? {} : { principalRefHash }),
    sourceHash,
  };
}

function projection(context: ActionContextDocument): Record<string, unknown> {
  return {
    agent_id: context.agent.agent_id,
    role: context.agent.role,
    privilege_scope: context.agent.privilege_scope,
    active_intents: context.agent.active_intents,
    action: context.action,
    run_id: context.run_id,
    session_id: context.session_id,
    thread_id: context.thread_id,
  };
}

export function validateDeferredChanges(params: {
  original_context: ActionContextDocument;
  current_context: ActionContextDocument;
  original_base: IntentBaseResult;
  current_base: IntentBaseResult;
  required_fields: string[];
}): void {
  const required = new Set(params.required_fields);
  const original = projection(params.original_context);
  const current = projection(params.current_context);
  const originalAction = { ...(original.action as Record<string, unknown>) };
  const currentAction = { ...(current.action as Record<string, unknown>) };
  if (required.has('action.target')) {
    delete originalAction.target;
    delete currentAction.target;
  }
  original.action = originalAction;
  current.action = currentAction;
  if (required.has('active_intents')) {
    delete original.active_intents;
    delete current.active_intents;
  }
  if (canonicalJsonForHash(original) !== canonicalJsonForHash(current)) {
    throw new Error('DEFER resolution changed fields outside required_fields');
  }
  if (required.has('policy_evaluation')) return;
  const allowedBase = new Set<string>();
  if (required.has('modified_arguments_hash')) allowedBase.add('modified_arguments_hash');
  for (const field of params.required_fields) {
    if (field.startsWith('approval_')) allowedBase.add(field);
  }
  const before = { ...params.original_base } as Record<string, unknown>;
  const after = { ...params.current_base } as Record<string, unknown>;
  for (const field of allowedBase) {
    delete before[field];
    delete after[field];
  }
  if (canonicalJsonForHash(before) !== canonicalJsonForHash(after)) {
    throw new Error('DEFER resolution changed base result outside required_fields');
  }
}
